linera-core/src/chain_worker/actor.rs

```diff
@@ -176,8 +176,45 @@ impl<StorageClient> ChainWorkerActor<StorageClient>
 where
     StorageClient: Storage + Clone + Send + Sync + 'static,
 {
-    /// Spawns a new task to run the [`ChainWorkerActor`], returning an endpoint for sending
-    /// requests to the worker.
+    /// Runs the [`ChainWorkerActor`], first by loading the chain state from `storage` then
+    /// handling all `incoming_requests` as they arrive.
+    ///
+    /// If loading the chain state fails the next request will receive the error reported by the
+    /// `storage`, and the actor will then try again to load the state.
+    pub async fn run(
+        config: ChainWorkerConfig,
+        storage: StorageClient,
+        executed_block_cache: Arc<ValueCache<CryptoHash, Hashed<Block>>>,
+        tracked_chains: Option<Arc<RwLock<HashSet<ChainId>>>>,
+        delivery_notifier: DeliveryNotifier,
+        chain_id: ChainId,
+        mut incoming_requests: mpsc::UnboundedReceiver<ChainWorkerRequest<StorageClient::Context>>,
+    ) {
+        let actor = loop {
+            let load_result = Self::load(
+                config.clone(),
+                storage.clone(),
+                executed_block_cache.clone(),
+                tracked_chains.clone(),
+                delivery_notifier.clone(),
+                chain_id,
+            )
+            .await;
+
+            match load_result {
+                Ok(actor) => break actor,
+                Err(error) => match incoming_requests.recv().await {
+                    Some(request) => request.send_error(error),
```

Comment:
What happens to failed requests?

Reply:
I believe it depends on the request, but most of them abort the operation :thinking:

Reply:
The current code will try to load a chain worker for every incoming request and fail them - one-by-one. I wonder whether the alternatives:
- failing all of them at once
- blocking until loaded
- panic(?)

were considered? Not being able to load a chain worker is a fatal error, right? It practically stalls the chain.

Reply:
I didn't give those alternatives much thought because I figured that a storage error could be temporary (e.g., network issue to the database). The option to block until the chain state is loaded is tricky, because of the mixture between `async` code (to load it) and synchronous code (the endpoint cache).